### PR TITLE
Fixes #9259 Change Tags APIs to conform with rest of the APIs

### LINKED ISSUE #9259: Change Tags APIs to conform with rest of the APIs
Tags APIs were coded long before other entities' APIs were added. Due to this, tag API does not follow the API convention that all the other entities are following. This issue makes **backward** incompatible changes to follow the same convention as glossaries.

## TagCategory API changes

**The name `TagCategory` is changing to `Classification` to adopt the more common terminology used in the industry**. The API endpoints are changed as follows:

|API functionality |Current API  | New API|
--- | --- | ---|
List tag categories|`GET .../tags` with no pagination|`GET .../classifications` that includes pagination|
Create a tag category|`POST .../tags`|`POST .../classifications`|
Get a tag category by name|`GET .../tags/{category}` that includes all the tags in the tag category|`GET .../classifications/name/{category}`. Children are no longer returned in the tag category. Use list tags APIs for this.|
Get a tag category by id|Not supported|`GET .../classifications/{id}`|
Update a tag category|`PUT .../tags/{category}|`PUT .../classifications`|
Delete a tag category|`DELETE .../tags/{id}|`DELETE .../classifications/{id}`|
Patch a tag category|Not supported|`PATCH .../classifications/{id}`|
Get versions of a tag category|Not supported|`GET .../classifications/{id}/versions`|
Get a specific version of a tag category|Not supported|`GET .../classifications/{id}/versions/{version}`|

## Tag API changes
Currently, only two levels of tag hierarchy are supported. The first level is called the primary tag and the next level is called the secondary tag. **We are moving away from this to any levels of nested tag hierarchy similar to Glossary and GlossaryTerms.**

|API functionality |Current API  | New API|
--- | --- | ---|
List tags|Not supported|`GET .../tags` that includes pagination and use filters to get tags under a tag category or a parent tag|
Create a tag|`POST .../tags/{category}` for the primary tag and `POST .../tags/{category}/{primaryTag} for the secondary tag`|`POST .../tags` with a field `parent` where parent tag term can be passed to create nested tag hierarchy|
Get a tag by name|`GET .../tags/{category}/{primaryTag}` for primary tag and `GET .../tags/{categoryName}/{primaryTag}/{secondaryTag}` for secondary tag|`GET .../tags/name/{fqn}`|
Get a tag by id|Not supported|`GET .../tags/{id}`|
Update a tag category|`PUT .../tags/{category}/{primaryTag}` to update primary tag and `PUT .../tags/{category}/{primaryTag}/secondaryTag` to update secondary tag|`PUT .../tags`|
Delete a tag|`DELETE .../{category}/{id}|`DELETE .../tags/{id}`|
Patch a tag category|Not supported|`PATCH .../tags/{id}.`** Use this for renaming the patch, changing hierarchy, etc.***|
Get versions of a tag category|Not supported|`GET .../tags/{id}/versions`|
Get a specific version of a tag category|Not supported|`GET .../tags/{id}/versions/{version}`|

## Implementation notes
The resource classes `TagResource` and `TagCategoryResource` use `EntityResource` as the base class to eliminate a lot of duplicate code. Similarly, `TagResourceTest` and `TagCategoryResourceTest` use `EntityResourceTest` as the base class to reuse code.

### PR DESCRIPTION
**Pull request by sureshms**
### Describe your changes :

See details in #9259 
#
### Type of change :

- [x] Bug fix
- [x] Improvement
- [x] New feature
- [x ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

See #9259 for details about backward incompatibility and how to migrate to the new set of tags APIs

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

### CODE REVIEW

File: openmetadata-service/src/main/java/org/openmetadata/service/util/FullyQualifiedName.java [hunk 1]
```diff
@@ -68,6 +68,15 @@ public static String getParent(String fqn) {
...
   }
 
+  public static String getRoot(String fqn) {
+    // Split fqn of format a.b.c.d and return the root a
+    String[] split = split(fqn);
+    if (split.length <= 1) {
+      return null;
+    }
+    return split[0];
+  }
```
Comment (nahuelverdugo) [inline openmetadata-service/src/main/java/org/openmetadata/service/util/FullyQualifiedName.java:78]: We are not using it anywhere.

**Comment by sureshms** [comment]
@chirag-madlani and @pmbrull, I have made changes in ingestion and UI code (to the extent I can). See comments TODO:9259 where the changes are still pending. The main thing that needs to be addressed is, Classification (previously called TagCategory) no longer has the field `children` to return tags. Instead, you have to list tags with filters (very similar to how it is done in Glossary and GlossaryTerms).

**Comment by ShaileshParmar11** [comment]
Hi, @sureshms I have pushed some UI changes when I was trying the `Create a classification` and `List classifications` API I got the below error.
```
{
    "responseMessage": "An exception with message [org.postgresql.util.PSQLException: ERROR: relation \"classification\" does not exist\n  Position: 44 [statement:\"/* ClassificationDAO.insert */ INSERT INTO classification (json) VALUES (:json :: jsonb)\", arguments:{positional:{1:{\"id\":\"a3c378b1-ad18-4e68-8837-785fcb562c62\",\"name\":\"string\",\"fullyQualifiedName\":\"string\",\"description\":\"string\",\"version\":0.1,\"updatedAt\":1671781657554,\"updatedBy\":\"admin\",\"deleted\":false,\"provider\":\"user\",\"mutuallyExclusive\":false}}, named:{json:{\"id\":\"a3c378b1-ad18-4e68-8837-785fcb562c62\",\"name\":\"string\",\"fullyQualifiedName\":\"string\",\"description\":\"string\",\"version\":0.1,\"updatedAt\":1671781657554,\"updatedBy\":\"admin\",\"deleted\":false,\"provider\":\"user\",\"mutuallyExclusive\":false}}, finder:[]}]] was thrown while processing request."
}
```

**Comment by pmbrull** [comment]
Thanks for handling the ingestion. Ping me if you need a hand

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/tags/TagResource.java [hunk 2]
```diff
@@ -47,538 +52,380 @@
...
 import org.openmetadata.service.util.ResultList;
 
 @Slf4j
 @Path("/v1/tags")
 @Api(value = "Tags resources collection", tags = "Tags resources collection")
 @Produces(MediaType.APPLICATION_JSON)
 @Consumes(MediaType.APPLICATION_JSON)
-@Collection(name = "tags", order = 6) // initialize after Glossary and GlossaryTerm
-public class TagResource {
+@Collection(name = "tags", order = 5) // initialize after Glossary and GlossaryTerm
```
[the remark this answers is not in the document]
Comment (sureshms) [reply openmetadata-service/src/main/java/org/openmetadata/service/resources/tags/TagResource.java:82]: The comments are wrong. The order of initialization is ClassificationResource (order=4), TagResource(5), Glossary(6), and GlossaryTerm(7). Fixing the comments.

File: openmetadata-service/src/main/java/org/openmetadata/service/util/FullyQualifiedName.java (hunk 1 above)
Comment (sureshms) [reply openmetadata-service/src/main/java/org/openmetadata/service/util/FullyQualifiedName.java:78]: Will keep it for now. I have plans of using this in a subsequent iteration.

**Comment by sureshms** [comment]
> Thanks for handling the ingestion. Ping me if you need a hand

@pmbrull all the required changes are not handled for ingestion. I could only go so far in renaming and making changes I could do.

See the earlier comment:
> I have made changes in ingestion and UI code (to the extent I can). See comments TODO:9259 where the changes are still pending. The main thing that needs to be addressed is, Classification (previously called TagCategory) no longer has the field children to return tags. Instead, you have to list tags with filters (very similar to how it is done in Glossary and GlossaryTerms).

**Comment by pmbrull** [comment]
> > Thanks for handling the ingestion. Ping me if you need a hand
> 
> @pmbrull all the required changes are not handled for ingestion. I could only go so far in renaming and making changes I could do.
> 
> See the earlier comment:
> 
> > I have made changes in ingestion and UI code (to the extent I can). See comments TODO:9259 where the changes are still pending. The main thing that needs to be addressed is, Classification (previously called TagCategory) no longer has the field children to return tags. Instead, you have to list tags with filters (very similar to how it is done in Glossary and GlossaryTerms).

perf, will push on this same branch

**Review by pmbrull — DISMISSED** [review]

**Review by pmbrull — DISMISSED** [review]
Thanks! LGTM

**Comment by pmbrull** [comment]
Cypress tags tests are failing though, we might need to review them

**Review by chirag-madlani — APPROVED** [review]
UI looks good too

**Comment by pmbrull** [comment]
> UI looks good too

@chirag-madlani don't we need to update cypress tests on Tags? All of them failing on MySQL and Postgres

**Review by pmbrull — APPROVED** [review]

**Comment by sureshms** [comment]
@chirag-madlani this PR is merged. Looking forward to you fix for cypress tests

**Pull request merged**